woocommerce/woocommerce-ios
# Edit Grouped Products: edit linked products (Part 2/3) - show a product list selector to add linked products

Product selector for grouped product's linked products #2199 

This PR implements the highlighted screen in the flow to edit linked products for a grouped product:

![this_pr](https://user-images.githubusercontent.com/1945542/87673125-99612300-c7a6-11ea-9b76-2f305d262e06.png)


## Changes

- Networking layer:
  - In `ProductsRemote.loadAllProducts`, added a parameter `excludedProductIDs: [Int64]` to exclude a given list of product IDs in the API call. Also updated the completion data type from `([Product]?, Error?)` to `Result<[Product], Error>`.
  - Updated unit tests:
    - Added the same `MockupNetwork` extension helper `pathComponents` that is exactly the same from `YosemiteTests`.
    - Added a test case to verify the new param in the API call from excluded product IDs is as expected
- Yosemite layer:
  - In `ProductAction.synchronizeProducts`, added two parameters `excludedProductIDs: [Int64] = []` and `shouldDeleteStoredProductsOnFirstPage: Bool = true`. In `ProductStore`'s handling of this action, `excludedProductIDs` is passed to `ProductsRemote` and `shouldDeleteStoredProductsOnFirstPage` is used to check whether to delete all the stored data when fetching the first page.
    - `shouldDeleteStoredProductsOnFirstPage` is needed because we don't want to delete all the stored products when fetching a list of products on the list selector screen
- App layer:
  - In `GroupedProductListSelectorDataSource`, added a new function `addProducts` that is called when the user adds products from the product list selector screen. Just to prevent from any duplicates due to `ResultsController` or any other mistakes, I also added a `Sequence` helper `removingDuplicates()` to remove all the duplicates after the first instance.
  - Prior to this product list selector screen, the cell always shows a checkmark in its accessory view if selected. However, we have a different selected UI in the product list selector screen so I removed the cell's accessory type configuration from `PaginatedListSelectorViewController` and implemented it in `PaginatedProductShippingClassListSelectorDataSource` and `ProductTaxClassListSelectorDataSource`
  - Created `ProductListSelectorViewController` that contains a child view controller `PaginatedListSelectorViewController` with data source `ProductListMultiSelectorDataSource` that contains an observable list of the latest product IDs
  - In `GroupedProductsViewController`, showed `ProductListSelectorViewController` and updated its data source on product list selector completion
  - Added/updated unit tests for the changes above

## Testing

- Go to the Products tab
- Tap on a grouped product
- Tap on the "Grouped products" row
- Tap on the "Add Products" button --> should see a list of products that **do not contain the pre-selected products plus the grouped product itself** (even though it's allowed to select itself in core), and the navigation bar has a search icon on the right
- Select a few products --> the navigation bar should indicate the number of selected products, and the search icon becomes "Done" button
- Tap "Done" --> the "Grouped products" row should show the updated total number of linked products
- Tap "Update" --> the linked products for the grouped product should be updated remotely

### Sanity check on affected screens

- Go to the Products tab
- Tap on a simple & physical product
- Tap on the price row
- Tap on the "Tax class" row --> should be able to select the tax class as before
- Go back to product details
- Tap on the shipping row
- Tap on the "Shipping class" row --> should be able to select the tax class as before

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/87673298-e5ac6300-c7a6-11ea-9808-1a514b319a54.gif)


state | light | dark
-- | -- | -- 
no products selected | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 44 00](https://user-images.githubusercontent.com/1945542/87672334-5eaabb00-c7a5-11ea-8d05-aaf5a31b0e1f.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 48 24](https://user-images.githubusercontent.com/1945542/87672548-be08cb00-c7a5-11ea-9e88-d0532ae5e582.png)
1 product selected | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 49 58](https://user-images.githubusercontent.com/1945542/87672737-f5777780-c7a5-11ea-8599-a446ade774f4.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 49 46](https://user-images.githubusercontent.com/1945542/87672721-f1e3f080-c7a5-11ea-9f74-abedbe4cbc17.png)
more than 1 products selected | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 44 34](https://user-images.githubusercontent.com/1945542/87672348-62d6d880-c7a5-11ea-9908-87db46d84832.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 20 45 14](https://user-images.githubusercontent.com/1945542/87672354-64080580-c7a5-11ea-9e5b-923454816de6.png)


Update release notes:
